### ISSUE TITLE
Add id's to expeditions in gamestate to simplify visualizer logic

### ISSUE THREAD

**Issue by wschella** [ZeusWPI/MOZAIC#44]
Currently, expeditions have no id's, because they don't need any from a bot or rules point of view, as there is no need for 'continuity' of one particular expedition. A step in the game is currently a valid begin state as well. This is not the case if we want to do fancy visualization though, as a visualizer would assume an expedition launched from some planet at some time, so that it can calculate the trajectory to it's destination, with each move being a _transition_ from the previous state, instead of a brand new location.

**Comment by iasoon** [comment]
Just putting another option on the board: if we log deltas instead of states, the visualizer could do things like this itself.

**Comment by tl3ilaxu** [comment]
Delas wouldn't really reduce complexity of the visualizer, if anything it would make it more complicated. The incoming data needs be processed more heavily and the current game state needs to be checked for validity. I like the idea of deltas though, it would reduce parsing times of otherwise huge gamestates with repeating data.

**Comment by iasoon** [comment]
@tl3ilaxu It would indeed add complexity. Checking is not needed though, that should already be done by the game server.
I think it matters a lot on the performance front, especially when you have lots of fleets. These will have to be repeated every step, versus a single dispatch entry in the delta case.
I agree that just dumping state is way easier and we should try that first, if only so that we have some example code for that style. However, I also root for a swift switch to deltas, and when we do, we should keep in mind that we don't need expedition ids :-)

**Comment by tl3ilaxu** [comment]
My problem with getting rid of id's is that the visualizer becomes too responsible for game logic. 
For example: we create a variation on planet wars where expeditions can be retreated back to their original planet. 
If we have id's we can just add a delta targeting that fleets id and turn it around. Without id's the game would need a different visualizer.
I am definitely for deltas but I think id's will give a stronger connection between gameserver and visualizer.

TLDR; I think keeping id's makes the visualizer more flexible, but deltas are bae

**Comment by wschella** [comment]
We can have both delta's and id's right? And there are many ways delta-states could be represented. We should take into considerations the limitations of delta-states though, in a high activity game, state will be changing in a rapid pace across the whole field, resulting in delta's almost as big as complete states.
To fix that, we would need various amounts of game-logic _in_ the visualizer, something we are trying to avoid.

As an example: expeditions change every turn, since the remaining_turns decreases. We could _not_ propagate these changes to the visualizer, as the next state is transparent. And on the other side of the spectrum: we propagate them every turn, resulting in delta's that offer no improvement.

Same goes for a planet's ship_count. 

Anyway, I think ID's are a must. Both from a logic and an efficiency point of view.

Sidenote: We could chose for a property scoped diff, instead of object scoped diff to battle the lack of efficiency in delta's.

**Comment by tl3ilaxu** [comment]
So consensus seems to be that we should add id's to the gamestate. This closes this issue. 
I think for now we keep the gamestates as they are and shelve the delta issue for the long term.
However we should create an issue discussing deltas an their specific implantation details.

**Comment by iasoon** [comment]
@tl3ilaxu unfortunately that does not close the issue, as our game state does not have ids!

**Comment by tl3ilaxu** [comment]
@iasoon I meant the discussion not the issue YOU NAZI
